primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,987 @@
+                let coord_left = Coord { height: height_child, idx: left_child_idx };
+                let left_child_hash = LazyBigMerkleTree::node(self, coord_left);
+
+                let right_child_idx = left_child_idx + 1;
+                let coord_right = Coord { height: height_child, idx: right_child_idx };
+                let right_child_hash = LazyBigMerkleTree::node(self, coord_right);
+
+                node_hash = merkle_tree::field_based_mht::smt::big_merkle_tree::BigMerkleTree::<F, T, P>::poseidon_hash(left_child_hash, right_child_hash);
+            }
+            return node_hash;
+        }
+
+        res.unwrap()
+    }
+
+    pub fn get_root(&self) -> F {
+        self.state.root.clone()
+    }
+
+    pub fn remove_node_from_cache(&mut self, coord: Coord) {
+        self.remove_from_cache(coord);
+    }
+
+    pub fn batch_poseidon_hash(input: Vec<F>) -> Vec<F> {
+        let output_vec = PoseidonBatchHash::<F, P>::batch_evaluate(&input);
+        output_vec.unwrap()
+    }
+
+    pub fn is_leaf_empty(&self, coord: Coord) -> bool {
+        // check that the index of the leaf is less than the width of the Merkle tree
+        assert!(coord.idx < self.state.width, "Leaf index out of bound.");
+        // check that the coordinates of the node corresponds to the leaf level
+        assert_eq!(coord.height, 0, "Coord of the node does not correspond to leaf level");
+
+        !self.state.present_node.contains(&coord)
+    }
+
+    pub fn process_leaves (&mut self, vec_leaf_op: Vec<OperationLeaf<F>>) -> F {
+
+        assert_eq!(T::MERKLE_ARITY, 2, "Arity of the Merkle tree is not 2.");
+
+        // Validate inputs
+        for i in 0..vec_leaf_op.len() {
+            // check that the index of the leaf to be inserted/removed is in range
+            assert!(vec_leaf_op[i].coord.idx < self.state.width, "Leaf index out of bound.");
+        }
+
+        // Mark nodes to recompute
+        let mut visited_nodes:HashSet<Coord> = HashSet::new();
+        let mut nodes_to_process_in_parallel:Vec<Vec<Coord>> = Vec::new();
+
+        // process the first level
+        let mut first_level_nodes:Vec<Coord> = Vec::new();
+        for j in 0..vec_leaf_op.len() {
+            let x = vec_leaf_op[j];
+            let coord = x.coord;
+            // visit parent
+            let height_parent = coord.height + 1;
+            let idx_parent = coord.idx / T::MERKLE_ARITY;
+            let parent_coord = Coord {height: height_parent, idx: idx_parent};
+            if !visited_nodes.contains(&parent_coord) {
+                // parent node not visited yet
+                visited_nodes.insert(parent_coord);
+                // insert node to process in parallel
+                first_level_nodes.push(parent_coord);
+            }
+        }
+        nodes_to_process_in_parallel.push(first_level_nodes);
+
+        // got to the upper levels until the root
+        let mut height = 1;
+        while height < self.height {
```

Comment:
Can't we merge this for with the previous "// process the first level" code block ?

Reply:
Cleaned up a little bit, but merging the two code blocks doesn't really improve readability, so it's best to keep it as it is.